File: app/static/base_types.ts
const MODEL_FIELDS = ["name", "base"]
const SUBMODEL_FIELDS = ["submodelStr", "seed", "batch", "learningRate"]

class Image {
    seed: number
    path: string
    imageset: ImageSet

    constructor(imageset: ImageSet, seed: number, path: string) {
        this.imageset = imageset
        this.seed = seed
        this.path = path
    }

    static from_json(imageset: ImageSet, input: any): Image {
        return new Image(imageset, input.seed, input.path)
    }
}

class ImageSet {
    prompt: string
    samplerStr: string
    cfg: number
    path: string
    key: string
    images: Array<Image>
    width: number
    height: number
    visible: boolean

    hide: boolean

    submodelSteps: SubModelSteps

    constructor(submodelSteps: SubModelSteps, prompt: string, samplerStr: string, cfg: number, width: number, height: number, path: string, key: string) {
        this.submodelSteps = submodelSteps
        this.prompt = prompt
        this.samplerStr = samplerStr
        this.cfg = cfg
        this.width = width
        this.height = height
        this.path = path
        this.key = key
        this.images = new Array()
        this.visible = false
        this.hide = false
    }

    resolution(): string {
        return `${this.width}x${this.height}`
    }

    static from_json(submodelSteps: SubModelSteps, input: any): ImageSet {
        const res = new ImageSet(submodelSteps, input.prompt, input.samplerStr, input.cfg, input.width, input.height, input.path, input.key)
        res.hide = input.hide
        for (const image of input.images) {
            res.images.push(Image.from_json(res, image))
        }
        return res
    }
}

class BaseModel {
    path: string
    visible: boolean
    constructor(path: string) {
        this.path = path
        this.visible = false
    }
}

class SubModelSteps extends BaseModel {
    submodel: SubModel
    steps: number
    imagesets: Array<ImageSet>
    canGenerate: boolean

    constructor(path: string, submodel: SubModel, steps: number, canGenerate: boolean) {
        super(path)
        this.submodel = submodel
        this.steps = steps
        this.imagesets = new Array()
        this.canGenerate = canGenerate
    }

    static from_json(submodel: SubModel, input: any): SubModelSteps {
        const res = new SubModelSteps(input.path, submodel, input.steps, input.canGenerate)
        for (const imageset of input.imageSets) {
            const oneIS = ImageSet.from_json(res, imageset)
            res.imagesets.push(oneIS)
        }
        return res
    }
}

class SubModel extends BaseModel {
    submodelStr: string
    seed: number
    submodelSteps: Array<SubModelSteps>
    batch: number
    learningRate: string
    extras: Set<string>
    model: Model
    canGenerate: boolean

    constructor(path: string, 
                 model: Model, submodelStr: string = "", seed: number = 0, batch: number = 1,
                 learningRate: string = "",
                 extras: Set<string> = new Set()) {
        super(path)
        this.visible = true
        this.model = model
        this.submodelStr = submodelStr
        this.seed = seed
        this.batch = batch
        this.learningRate = learningRate
        this.extras = extras
        this.submodelSteps = []
        this.canGenerate = false
    }

    static from_json(model: Model, input: any): SubModel {
        const res = new SubModel(input.path, model)
        res.submodelStr = input.submodelStr
        res.seed = input.seed
        res.batch = input.batch
        res.learningRate = input.learningRate
        res.extras = new Set(input.extras)

        for (const oneSteps of input.submodelSteps) {
            const submodelSteps = SubModelSteps.from_json(res, oneSteps)
            res.submodelSteps.push(submodelSteps)
            if (submodelSteps.canGenerate) {
                res.canGenerate = true
            }
        }
        return res
    }
}

class Model extends BaseModel {
    name: string
    base: string
    submodels: Array<SubModel>
    canGenerate: boolean

    constructor(path: string, name: string, base: string) {
        super(path)
        this.visible = true
        this.path = path
        this.name = name
        this.base = base
        this.submodels = new Array()
        this.canGenerate = false
    }

    static from_json(input: any): Model {
        const res = new Model(input.path, input.name, input.base)
        for (const submodelIn of input.submodels) {
            const submodel = SubModel.from_json(res, submodelIn)
            res.submodels.push(submodel)
            if (submodel.canGenerate) {
                res.canGenerate = true
            }
        }
        return res
    }
}
    
export { SubModelSteps, SubModel, Model, BaseModel, ImageSet, Image,
         MODEL_FIELDS, SUBMODEL_FIELDS }

File: app/base_types.py
import json
from pathlib import Path
from typing import Iterable, Dict, Set, List
import inspect

class BaseModel:
    def to_dict(self, attributes: Iterable[str] = None) -> Dict[str, any]:
        if attributes is None:
            attributes = sorted(self.__dict__.keys()) 

        res: Dict[str, any] = {}
        for attr in attributes:
            value = getattr(self, attr)
            if isinstance(value, set):
                value = list(value)
            if isinstance(value, Iterable) and len(value) > 0 and isinstance(value[0], BaseModel):
                value = [child.to_dict() for child in value]
            if not isinstance(value, list) and not isinstance(value, str) and not isinstance(value, int) and not isinstance(value, float) and not isinstance(value, bool):
                if attr == "model_path":
                    continue
                print(f"skip attribute {attr} value {value}")
                continue
            res[attr] = value
        
        res["key"] = self.get_key()
        return res
    
    def get_key(self, attributes: List[str] = None) -> str:
        res: List[str] = []
        if attributes is None:
            attributes = sorted(self.__dict__.keys())
        for attr in attributes:
            value = getattr(self, attr)

            if isinstance(value, int) or isinstance(value, str):
                res.append(f"{attr}={value}")
            elif isinstance(value, set):
                values: List[str] = []
                for item in sorted(list(value)):
                    values.append(str(item))
                values_str = ",".join(values)
                res.append(f"{attr}=({values_str})")
        
        return ",".join(res)

class ImageSet: pass
class Model: pass
class SubModel: pass

class SubModelSteps(BaseModel):
    steps: int
    imageSets: List[ImageSet]
    submodel: SubModel
    canGenerate: bool
    model_path: Path

    def __init__(self, submodel: SubModel, steps: int, canGenerate: bool = False, model_path: Path = None):
        self.submodel = submodel
        self.steps = steps
        self.imageSets = list()
        self.canGenerate = canGenerate
        self.model_path = model_path
    
    def image_path(self) -> Path:
        return Path(self.submodel.image_path(), self.get_key())

    def get_key(self) -> str:
        return f"steps={self.steps}"

    def to_dict(self) -> Dict[str, any]:
        attributes = set(self.__dict__.keys())
        attributes.remove("submodel")
        res = super().to_dict(sorted(attributes))
        res['path'] = str(self.image_path())
        return res

class SubModel(BaseModel):
    submodelStr: str
    seed: int
    batch: int
    learningRate: str
    submodelSteps: List[SubModelSteps]
    extras: Set[str]
    model: Model

    def __init__(self, model: Model = None, submodelStr = "", seed = 0, batch = 1, learningRate = "", extras: Set[str] = set()):
        self.model = model
        self.submodelStr = submodelStr
        self.seed = seed
        self.batch = batch
        self.learningRate = learningRate
        self.extras = extras
        self.submodelSteps = list()

    def get_key(self) -> str:
        res = []
        res.append(f"seed={self.seed}")
        res.append(f"batch={self.batch}")
        res.append(f"LR={self.learningRate}")
        if self.extras:
            # extras come at end.
            res.append(",".join(sorted(list(self.extras))))
        return ",".join(res)

    def image_path(self) -> Path:
        return Path(self.model.image_path(), self.get_key())

    def to_dict(self) -> Dict[str, any]:
        attributes = set(self.__dict__.keys())
        attributes.remove("model")
        res = super().to_dict(sorted(attributes))
        res['path'] = str(self.image_path())
        return res

class Model(BaseModel):
    name: str
    base: str
    submodels: List[SubModel]

    def __init__(self, name = "", base = ""):
        self.name = name
        self.base = base
        self.submodels = list()
    
    def get_key(self) -> str:
        res = self.name
        if self.base:
            res += f"+{self.base}"
        return res

    def image_path(self) -> Path:
        return Path(self.get_key())

    def to_dict(self) -> Dict[str, any]:
        res = super().to_dict()
        res['path'] = str(self.image_path())
        return res
    
class Image(BaseModel):
    imageset: ImageSet
    seed: int

    def __init__(self, imageset: ImageSet, seed: int):
        self.imageset = imageset
        self.seed = seed

    def path(self) -> Path:
        png = f"{self.seed:010}.png"
        return Path(self.imageset.path(), png)

    def to_dict(self) -> Dict[str, any]:
        attributes = set(self.__dict__.keys())
        attributes.remove("imageset")
        res = super().to_dict(sorted(attributes))
        res['path'] = str(self.path())
        return res
    
class ImageSet(BaseModel):
    model: Model
    submodel: SubModel
    submodelSteps: SubModelSteps
    samplerStr: str
    cfg: int
    prompt: str
    images: List[Image]

    def __init__(self, model: Model, submodel: SubModel, submodelSteps: SubModelSteps, prompt: str, samplerStr: str, cfg: int, seeds: Iterable[int] = []):
        self.model = model
        self.submodel = submodel
        self.submodelSteps = submodelSteps
        self.prompt = prompt
        self.samplerStr = samplerStr
        self.cfg = cfg

        for seed in seeds:
            self.images.append(Image(self, seed))
        self.images = list()

    def get_key(self) -> str:
        return super().get_key(["prompt", "samplerStr", "cfg"])

    def path(self) -> Path:
        endStr = f"sampler={self.samplerStr},cfg={self.cfg}"
        return Path(self.model.get_key(), self.submodel.get_key(), self.submodelSteps.get_key(), self.prompt, endStr)

    def to_dict(self) -> Dict[str, any]:
        attributes = set(self.__dict__.keys())
        attributes.remove("model")
        attributes.remove("submodel")
        attributes.remove("submodelSteps")
        res = super().to_dict(sorted(attributes))
        res['path'] = str(self.path())
        return res
